File: daily_problems/2024/06/0617/personal_submission/cf535d_yawn_sean.cpp
#include <bits/stdc++.h>
#pragma GCC optimize("O3")
using namespace std;

signed main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    int n, m;
    cin >> n >> m;

    string s;
    cin >> s;
    int len = s.size();

    vector<int> pos(m);
    for (int i = 0; i < m; i ++)
        cin >> pos[i], pos[i] --;
    
    vector<int> z(len);

    int j = 0;
    for (int i = 1; i < len; i ++) {
        z[i] = (j + z[j] <= i ? 0 : min(j + z[j] - i, z[i-j]));
        while ((i + z[i] < len) && (s[z[i]] == s[i+z[i]])) z[i] ++;
        if (j + z[j] < i + z[i]) j = i; 
    }

    for (int i = 1; i < m; i ++) {
        int idx = pos[i] - pos[i-1]; 
        if (idx >= len) continue;
        if (z[idx] + idx != len) {
            cout << "0\n";
            return 0;
        }
    }

    vector<int> diff(n + 1);
    for (auto &p: pos)
        diff[p] ++, diff[p+len] --;

    long long ans = 1;
    int mod = 1e9 + 7;
    for (int i = 0; i < n; i ++) {
        diff[i+1] += diff[i];
        if (!diff[i]) ans = ans * 26 % mod;
    }
    cout << ans << '\n';

    return 0;
}

File: daily_problems/2024/06/0617/personal_submission/cf535d_yawn_sean.py
# Submission link: https://codeforces.com/contest/535/submission/266124866
def z_algorithm(s):

    if isinstance(s, str):
        s = [ord(c) for c in s]

    n = len(s)
    if n == 0:
        return []

    z = [0] * n
    j = 0
    for i in range(1, n):
        z[i] = 0 if j + z[j] <= i else min(j + z[j] - i, z[i - j])
        while i + z[i] < n and s[z[i]] == s[i + z[i]]:
            z[i] += 1
        if j + z[j] < i + z[i]:
            j = i
    z[0] = n

    return z

def main():
    n, m = MII()
    s = I()
    pos = LGMI()

    k = len(s)
    z = z_algorithm(s)

    for i in range(m - 1):
        if pos[i + 1] - pos[i] >= k: continue
        j = pos[i + 1] - pos[i]
        if z[j] + j != k: exit(print(0))

    diff = [0] * (n + 1)
    for p in pos:
        diff[p] += 1
        diff[p+k] -= 1

    ans = 1
    mod = 10 ** 9 + 7

    for i in range(n):
        diff[i+1] += diff[i]
        if diff[i] == 0:
            ans *= 26
            ans %= mod

    print(ans)